File: solutions/php/food-chain/1/FoodChain.php
<?php

/*
 * By adding type hints and enabling strict type checking, code can become
 * easier to read, self-documenting and reduce the number of potential bugs.
 * By default, type declarations are non-strict, which means they will attempt
 * to change the original type to match the type specified by the
 * type-declaration.
 *
 * In other words, if you pass a string to a function requiring a float,
 * it will attempt to convert the string value to a float.
 *
 * To enable strict mode, a single declare directive must be placed at the top
 * of the file.
 * This means that the strictness of typing is configured on a per-file basis.
 * This directive not only affects the type declarations of parameters, but also
 * a function's return type.
 *
 * For more info review the Concept on strict type checking in the PHP track
 * <link>.
 *
 * To disable strict typing, comment out the directive below.
 */

declare(strict_types=1);

class FoodChain
{
    private array $animals = [
        'fly',
        'spider',
        'bird',
        'cat',
        'dog',
        'goat',
        'cow',
        'horse'
    ];
    
    private array $reactions = [
        'fly' => "I don't know why she swallowed the fly. Perhaps she'll die.",
        'spider' => "It wriggled and jiggled and tickled inside her.",
        'bird' => "How absurd to swallow a bird!",
        'cat' => "Imagine that, to swallow a cat!",
        'dog' => "What a hog, to swallow a dog!",
        'goat' => "Just opened her throat and swallowed a goat!",
        'cow' => "I don't know how she swallowed a cow!",
        'horse' => "She's dead, of course!"
    ];
    
    public function verse(int $verseNumber): array
    {
        $index = $verseNumber - 1;
        $animal = $this->animals[$index];
        $lines = [];
        
        // First line
        $lines[] = "I know an old lady who swallowed a $animal.";
        
        // Special case for horse (last verse) - ends immediately
        if ($animal === 'horse') {
            $lines[] = $this->reactions[$animal];
            return $lines;
        }
        
        // Reaction line specific to the animal
        $lines[] = $this->reactions[$animal];
        
        // Cumulative lines - going backwards from current animal down to fly
        for ($i = $index; $i > 0; $i--) {
            $current = $this->animals[$i];
            $previous = $this->animals[$i - 1];
            
            $line = "She swallowed the $current to catch the $previous";
            
            // Special case: spider has extra description when being caught
            if ($previous === 'spider') {
                $line .= " that wriggled and jiggled and tickled inside her";
            }
            
            $line .= ".";
            $lines[] = $line;
        }
        
        // Final fly line (only add if this isn't already the fly verse)
        if ($animal !== 'fly') {
            $lines[] = $this->reactions['fly'];
        }
        
        return $lines;
    }
    
    public function verses(int $start, int $end): array
    {
        $result = [];
        
        for ($i = $start; $i <= $end; $i++) {
            $result = array_merge($result, $this->verse($i));
            
            // Add empty line between verses (but not after the last one)
            if ($i < $end) {
                $result[] = '';
            }
        }
        
        return $result;
    }
    
    public function song(): array
    {
        return $this->verses(1, count($this->animals));
    }
}

File: solutions/java/food-chain/1/src/main/java/FoodChain.java
class FoodChain {
    private static final String[] ANIMALS = {
        "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"
    };

    private static final String[] REACTIONS = {
        "",
        "It wriggled and jiggled and tickled inside her.",
        "How absurd to swallow a bird!",
        "Imagine that, to swallow a cat!",
        "What a hog, to swallow a dog!",
        "Just opened her throat and swallowed a goat!",
        "I don't know how she swallowed a cow!",
        "She's dead, of course!"
    };

    String verse(int verse) {
        int idx = verse - 1;
        StringBuilder sb = new StringBuilder();
        
        // First line
        sb.append("I know an old lady who swallowed a ").append(ANIMALS[idx]).append(".");
        
        // The horse is a special case
        if (idx == 7) {
            sb.append("\n").append(REACTIONS[idx]);
            return sb.toString();
        }
        
        // Add the unique reaction line for verses 2-7
        if (idx > 0) {
            sb.append("\n").append(REACTIONS[idx]);
        }
        
        // Build the cumulative "swallowed to catch" chain backwards
        for (int i = idx; i > 0; i--) {
            sb.append("\nShe swallowed the ").append(ANIMALS[i])
              .append(" to catch the ").append(ANIMALS[i - 1]);
            
            // The spider has a special extension when it's the one being caught
            if (i - 1 == 1) {
                sb.append(" that wriggled and jiggled and tickled inside her");
            }
            sb.append(".");
        }
        
        // The final line for all non-horse verses
        sb.append("\nI don't know why she swallowed the fly. Perhaps she'll die.");
        return sb.toString();
    }

    String verses(int startVerse, int endVerse) {
        StringBuilder sb = new StringBuilder();
        for (int i = startVerse; i <= endVerse; i++) {
            sb.append(verse(i));
            // Add a blank line between verses, but not after the final verse
            if (i < endVerse) {
                sb.append("\n\n");
            }
        }
        return sb.toString();
    }
}